libflux/src/ast/walk/mod.rs

```diff
@@ -231,109 +231,114 @@ impl<'a> Node<'a> {
 // Done is called on that Node to signal that we are done with that Node.
 //
 // If Visit returns None, walk will not recurse on the children.
-pub trait Visitor: Sized {
+//
+// Note: the Rc in visit is to allow for multiple ownership of a node, i.e.
+//       a visitor can own a node as well as the walk funciton. This allows
+//       for nodes to persist outside the scope of the walk function and to
+//       be cleaned up once all owners have let go of the reference.
+pub trait Visitor<'b>: Sized {
     // Visit is called for a node.
     // The returned visitor will be used to walk children of the node.
     // If visit returns None, walk will not recurse on the children.
-    fn visit(&self, node: &Node) -> Option<Self>;
+    fn visit<'a>(&self, node: Rc<Node<'b>>) -> Option<Self>;
     // Done is called for a node once it has been visited along with all of its children.
-    fn done(&self, _: &Node) {} // default is to do nothing
+    fn done(&self, _: Rc<Node>) {} // default is to do nothing
 }
 
 // Walk recursively visits children of a node.
 // Nodes are visited in depth-first order.
-pub fn walk<T>(v: &T, node: Node)
+pub fn walk<'a, T>(v: &T, node: Rc<Node<'a>>)
 where
-    T: Visitor,
+    T: Visitor<'a>,
 {
```

Comment:
Should we create separate public/private functions to encapsulate the creation of the Rc for the first node?

Something like:

```
pub fn walk<T>(v: &T, node: Node<'a>)
where
    T: Visitor<'a>,
{
    walk_rec(visitor, Rc::new(node))
}


fn walk_rec<'a, T>(v: &T, node: Rc<Node<'a>>) 
where
    T: Visitor<'a>,
{
   // normal recursive walk function definition
}

```

Thoughts?

Reply:
Would the private method then recurse to the public on? If that's the case we could remove all the Rc creation in walk_rec.